source/dataman/CLS/CLSDbUpgrade1Pt2.cpp

```diff
@@ -27,84 +28,132 @@ bool CLSDbUpgrade1Pt2::upgradeNecessary() const
-	QMap<int, QString> facilityTableThumbnails;
-
-	QSqlQuery thumbnailTableQuery = databaseToUpgrade_->select(thumbnailTableName, "id, title", QString("objectTableName=%1").arg("AMFacility_table") );
-	if (thumbnailTableQuery.exec()) {
-		while (thumbnailTableQuery.next()) {
-			facilityTableThumbnails.insert(thumbnailTableQuery.value(0).toInt(), thumbnailTableQuery.value(1).toString()); //id, title
-//			qDebug() << "==== " << thumbnailTableQuery.value(0).toInt() << thumbnailTableQuery.value(1).toString();
-		}
-	}
-	thumbnailTableQuery.finish();
-
-	bool correctThumbnailTableData = (facilityTableThumbnails.count() == 1) && (facilityTableThumbnails.values().at(0) == targetFacilityName_);
-//	qDebug() << "==== " << correctThumbnailTableData;
+	QMap<int, QString> facilityTableThumbnails = queryAMFacilityTableThumbnails();
 
-	return !correctThumbnailTableData;
+	return facilityTableThumbnails.count() != 1 || facilityTableThumbnails.values().at(0) != targetFacilityName_;
 }
 
 bool CLSDbUpgrade1Pt2::upgradeImplementation()
 {
 	int dbResult = -1;
 
-	// Prologue.
+	// Prologue:
+	//          - Start transation
+	//          - Get the current thumbnails for AMFacility_table
 	////////////////////////////////////////////////
 
 	databaseToUpgrade_->startTransaction();
 
-	QString thumbnailTableName = AMDbObjectSupport::thumbnailTableName();
+	QMap<int, QString> facilityTableThumbnails = queryAMFacilityTableThumbnails();
+	if (facilityTableThumbnails.count() == 1 && facilityTableThumbnails.values().at(0) == targetFacilityName_) {
+		databaseToUpgrade_->rollbackTransaction();
+		AMErrorMon::alert(this, CLSDbUpgrade1Pt2_ALREADY_UPDATED, QString("The thumbnail table has the correct data. No need to updgrade."));
+		return true;
+	}
 
-	// Stage 1: update the thumbnail table if required
-	//   - The AMDbObjectThumbnails_table doesn't have thumbnail for the given beamline name
-	QMap<int, QString> facilityTableThumbnails;
-	QSqlQuery thumbnailTableQuery = databaseToUpgrade_->select(thumbnailTableName, "id, title", QString("objectTableName=%1").arg("AMFacility_table") );
-	if (thumbnailTableQuery.exec()) {
-		while (thumbnailTableQuery.next()) {
-			facilityTableThumbnails.insert(thumbnailTableQuery.value(0).toInt(), thumbnailTableQuery.value(1).toString()); //id, title
+	// stage 1:
+	//         - update the thumbnail table
+	//           * remove the unnecessary thumbnail data for AMFacility table
+	//
+	foreach (int id, facilityTableThumbnails.keys()) {
+		if (facilityTableThumbnails.value(id) != targetFacilityName_) {
+			dbResult = databaseToUpgrade_->deleteRow(id, thumbnailTableName_);
+			if ( dbResult == 0 ){
+				AMErrorMon::alert(this, CLSDbUpgrade1Pt2_FAIL_TO_DELETE_FACILITY_THUMBNAIL, QString("Could not delete the thumbnail (%1) in %2").arg(id).arg(thumbnailTableName_));
+				databaseToUpgrade_->rollbackTransaction();
+				return false;
+			}
+
+			facilityTableThumbnails.remove(id);
```

Comment:
I find that removing elements from a list that is being used in the for loop arguments can lead to some unpredictable behaviour.  I would suggest something like:

```c++
QList<int> idsToRemove;
foreach(){
    if(){
        // stuff
        idsToRemove << id;
    }
}
foreach(id, idsToRemove)
    facilityTableThumbnails.remove(id);
```